Transfinite interpolation manifold implementation

kronbichler opened the pull request:
This work is the first step to implement a manifold that does transfinite interpolation over a larger range of elements. As the class is initialized now, it takes a cell (on a coarse mesh) over which the transfinite interpolation is defined. I do not really like this interface because in case there are many cells that would need this treatment, one needs a separate manifold for each segment. @luca-heltai what is your opinion on the interface and how to best make use of this?

I could imagine that we make this the default description when a user set a boundary manifold but not an interior one. Or we let a user set this mapping as an option. This would invoke some logic when we need to identify the appropriate mother cell but it would otherwise be much cleaner. Any opinions on this?

If we delete the `Boundary` classes (#2418), the work would not be too tremendous. In addition, such an approach would avoid the problem from #3874 by construction.

This PR is work in progress: 3D is not tested at all (and I am pretty sure it is wrong) and additionally one might like to derive the class not from `ChartManifold` but rather implement `add_new_points` because that would be likely more efficient.

bangerth commented:
Yes, setting a manifold for (coarse) cells individually seems undesirable. At least we had the foresight to make `types::manifold_id` an `unsigned int` (as opposed to the `unsigned char` for `types::boundary_id`).

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,60 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * mesh according to a curved boundary description is applied to more than one
+ * element layer once the mesh gets refined, which is necessary to get
+ * high-order convergence rates (without this class, the error on cells close
+ * to the boundary does not exceed 3 on a circle).
```
error *order*?

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,60 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * interpolation. As opposed to setting the manifold description only on the
+ * boundary of a mesh, this approach makes sure that the deformation of the
+ * mesh according to a curved boundary description is applied to more than one
+ * element layer once the mesh gets refined, which is necessary to get
```
this is more than one *fine* mesh layer, right? But only one *coarse* mesh layer.

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,60 @@ class TorusManifold : public ChartManifold<dim,3,3>
+/**
+ * A mapping class that smoothly transitions from different manifold
+ * descriptions on different faces by a so-called transfinite
+ * interpolation. As opposed to setting the manifold description only on the
```
If I understand this correctly, what you're doing is taking a curved description for one or more faces of a cell, and then construct some kind of smooth continuation of this description into the interior. Is this correct? If so, would you add a sentence or two here that explains how this is actually done?

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,60 @@ class TorusManifold : public ChartManifold<dim,3,3>
+class TransfiniteInterpolationManifold : public ChartManifold<dim,spacedim,dim>
+{
+public:
+  void initialize (const typename Triangulation<dim,spacedim>::cell_iterator &cell);
```
you can only set a manifold once you have a mesh, so shouldn't it be possible to move the initialization with a cell into the constructor of the class?

bangerth replied:
I think you should also mention that the function evaluates the vertices and manifolds *at the time this function is called with a cell*, rather than saving the cell iterator and looking things up later on. This, for example, doesn't allow for meshes that change, and requires setting manifold ids on the faces *before* calling this function.

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,60 @@ class TorusManifold : public ChartManifold<dim,3,3>
+  /**
+   * The manifold descriptions for the surrounding faces.
+   */
+  const Manifold<dim,spacedim> *face_manifolds[GeometryInfo<dim>::faces_per_cell];
```
what do you do about edges in 3d?

bangerth replied:
Also, would it be better to use `SmartPointer` to make sure the object pointed to doesn't go away?

kronbichler replied:
Sure, I will. By the new design I will automatically get the right edges in 3D.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -549,6 +549,127 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+  // check that the point is in the unit interval which is the chart that is
+  // prescribed
+  for (unsigned int d=0; d<dim; ++d)
+    Assert(chart_point[d] >= -1e-7 && chart_point[d] <= 1.+1e-7,
```
Don't we have a tool function for this in `GeometryInfo`?

bangerth replied:
In fact, you even use it further down: `GeometryInfo<dim>::is_inside_unit_cell`

kronbichler replied:
I will use that in my modified version.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -549,6 +549,127 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+                      + std::to_string(d)));
+
+  // formula see wikipedia
+  // https://en.wikipedia.org/wiki/Transfinite_interpolation
```
Maybe put this link somewhere in the documentation?

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -549,6 +549,127 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+template <int dim, int spacedim>
+DerivativeForm<1,dim,spacedim>
+TransfiniteInterpolationManifold<dim,spacedim>
+::push_forward_gradient(const Point<dim> &chart_point) const
```
could this function be more accurately implemented if the manifolds that describe the faces happen to also be `ChartManifold`-derived classes for which we can call `push_forward_gradient()`?

kronbichler replied:
I think so, yes. But I will now go away from the `ChartManifold` anyway.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -549,6 +549,127 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+        modified[d] += eps;
+      Tensor<1,spacedim> difference = push_forward(modified) - point;
+      for (unsigned int e=0; e<spacedim; ++e)
+        grad[e][d] = difference[e] / (modified[d] - chart_point[d]);
```
not that it will make much of a difference, but you know what `modified[d] - chart_point[d]` is (namely +eps or -eps). It may be easier to read if you used it that way.

kronbichler commented:
My new idea for a better interface for this class is to attach a `TransfiniteInterpolationManifold` with a coarse grid on a Triangulation. This way, we only need to put a single manifold class to a grid and it can transparently handle all boundaries. I will try to design a new implementation during the coming days that goes along those lines. It will not be a `ChartManifold` because the resulting manifold class is an atlas of charts rather than a single chart, but we really only need the `get_new_point` and `add_new_points` interfaces which is possible.

bangerth commented:
Having an atlas sounds a lot like having a chart for each cell :-)

If you think you can find a good interface, I'm curious. But in hindsight, I think it is also ok to do as you originally suggested -- it is really only one manifold per *coarse* mesh cell if done correctly. We can probably deal with this.

kronbichler commented:
> Having an atlas sounds a lot like having a chart for each cell :-)

I want to define it for each coarse cell, before the refinement is started. That would be the main difference. Anyway, I need to experiment with the implementation to see how easy or difficult it is in the end.

bangerth commented:
I see. You'll define the atlas for all coarse mesh cells, but it will of course only be used for those cells that have the right manifold id set. That seems sensible.

kronbichler commented:
Now I implemented a basic version of what I had in mind. Here is an example mesh that is generated in the (WIP) test I added right now:
![nice_grid](https://cloud.githubusercontent.com/assets/8276200/26563307/7ba26eca-44d0-11e7-9183-a67aa220ee8e.png)

Note how nicely the outer circular shape is transformed into a straight line at the inner circle boundary.

The current code is still WIP as it is untested in 3D, but the basic concept should be clear.

Since setting the new manifold is a bit tricky: Since the transfinite manifold uses the surrounding entities of a cell, I cannot currently use it in 3D where I would need the faces to also deform according to a transfinite description, see the approach applied in #3886. I still do not have a good concept for that, so I am happy for suggestions.

As can be seen by the current class design, the transfinite interpolations is tightly coupled to the underlying triangulation. It appears that the transfinite manifold should pertain to the triangulation it lives on and be the new default manifold description of a triangulation rather than the current `StraightBoundary`. (I will need to build in a few tricks to make it really efficient in that case, though.)

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,150 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * manifold is later attached to the same triangulation.
+   *
+   * @note For performance and accuracy reasons, it is recommended to apply
+   * the transfinite manifold to as coarse meshes as possible. Regarding accurcay, the curved
```
-> as a coarse a mesh as possible

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,150 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * @note For performance and accuracy reasons, it is recommended to apply
+   * the transfinite manifold to as coarse meshes as possible. Regarding accurcay, the curved
+   * description can only be applied to new points created from a given
+   * surrounding, and the grid quality is typically higher when extending the
```
surrounding -> neighborhood?

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,150 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * weights for the surrounding points according to which the manifold
+   * determines the new point's position.
+   *
+   * The implementation in this class overrides the method in the base class
```
I don't think the base class actually has an implementation. Do you mean "This function implements the corresponding interface of the base class..."?

kronbichler replied:
It actually has, see `manifold.cc:87`. The implementation calls `get_intermediate_point` which in turn calls `project_to_manifold`. The only function without implementation is `project_to_manifold`, but we bypass that function here and I would also not know what to do at this point.

bangerth replied:
Oh, I see.

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,150 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * @p new_points equals the size at entry plus the number of rows in
+   * @p weights.
+   *
+   * The implementation in this class overrides the method in the base class
```
same as above

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -106,7 +106,7 @@ namespace internal
         // exit if the point is not in the cell (this is the only case where the
         // discriminant is negative)
-        if (discriminant < 0.0)
+        if (discriminant < 0.0 || ( a == 0.0 && b == 0.0))
```
the changes in this file would also have been appropriate for a separate PR. Out of curiosity, would it not have been enough to test for `discriminant <= 0.0`? Or can you think of *legitimate* cases where `discriminant==0` but `!(a==0 && b == 0)`?

kronbichler replied:
I have opened #4450 where this one is addressed. Indeed the case I wanted to catch was `a==0` and `b==0` in which case the discriminant is always zero, so your suggestion works just fine.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -549,6 +551,284 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+  return best_fit;
+
+  // To each surrounding point, find the closest vertex in the cell
+  /*
```
do you still need this block?

kronbichler replied:
No, it will go away.

bangerth commented on source/grid/manifold_lib.cc:
```diff
@@ -549,6 +551,284 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+  for (unsigned int i=0; i<surrounding_points.size(); ++i)
+    chart_points[i] = pull_back(cell, surrounding_points[i]);
+
+  FlatManifold<dim> sub_manifold;
```
Might make it `static const`?

Same elsewhere as appropriate.

kronbichler replied:
But `FlatManifold` is for free to set up, so I do not want to clutter our already large set of static variables. I will make `MappingQ1` that I also use a member variable and could do the same for `FlatManifold` if you find it nicer.

bangerth replied:
Yes, maybe so.

My worries when I see this is always "what happens if `FlatManifold` suddenly acquires state and becomes more expensive?". That's unlikely to happen for `FlatManifold`, but we've come up with lots of things that had simple implementations in the beginning and then later on become significantly more complicated. We can guard against that by making sure we don't create too many *potentially* expensive objects here and there.

As I said, it's not likely going to be a problem here. Do as you see fit.

bangerth commented on tests/manifold/transfinite_manifold_01.cc:
```diff
@@ -0,0 +1,68 @@
+      bool cell_at_boundary = false;
+      for (unsigned int face=0; face<GeometryInfo<dim>::faces_per_cell; ++face)
+        if (cell->at_boundary(face))
+          cell_at_boundary = true;
```
You can just ask `cell->at_boundary()`, which does this loop for you.

kronbichler replied:
Thanks ;-)

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +503,150 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   */
+  Point<dim>
+  pull_back(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+            const Point<spacedim> &p) const;
```
Since you are, after all and in some sense, implementing the functions of `ChartManifold`, why not derive the class from `ChartManifold` and implement that class's functions in terms of the ones you already have here plus `get_cell_around_points()`?

On a side note, this function may as well be `static` since it doesn't access anything it can't get through

kronbichler replied:
I don't think we can derive from a `ChartManifold`. While I could call `pull_back` that simply finds the point around the ones given as you suggest, I can not reasonably define the `push_forward` function because that function **must** know which of the charts to pick from the atlas. Just from the unit coordinates there is no way to tell. Unless I'm overlooking something here, I do not see why to expose those functions or wrap them into the interface of `ChartManifold`.

kronbichler replied:
Yes, that's true and we might add that if we see a need for it in the future. But it is a bit tricky to get the connectivity right (I would shift the x coordinate by 2 between the different points) so I'm not doing it right now.

kronbichler commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,182 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * manifold is later attached to the same triangulation.
+   *
+   * @note For performance and accuracy reasons, it is recommended to apply the
+   * transfinite manifold to as coarse a mesh as possible. Regarding accurcay,
```
I have a problem with this recommendation: Looking at cylindrical part of the figure (from #4451) towards the middle:
![grid_cylinder](https://cloud.githubusercontent.com/assets/8276200/26629673/b6b12202-4603-11e7-9d27-04c6bee17b5d.png)
shows that some cells are strange/inverted. This is because the detection of the coarse mesh cell fails: Two points along the circular shape end up outside the domain spanned by `MappingQ1` that I use for the identification of the coarse cell, and they get closer to the neighbor. It can be fixed by going to a mesh one level finer:
![grid_cylinder_2](https://cloud.githubusercontent.com/assets/8276200/26630576/a12d04c4-4607-11e7-8880-428086fc4d25.png)

Thus, we should probably recommend to try as coarse mesh as possible but if one detects problems initialize again on a mesh one level finer.

It would of course have been even better if we weren't throwing away the context of the current mesh cell in `get_new_point` or `add_new_points` by passing the cell where we actually want to compute the new points. This avoids doing a manual search which might get things wrong since I'd only need to go to the appropriate parent. But changing interface because of this one class is not what I want to do.

bangerth replied:
Yes, indeed. We also always thought of manifolds as independent of a mesh, just describing the geometry. Part of the appeal of the `Manifold` class over the `Boundary` class is that it does not depend on triangulations.

kronbichler commented:
This should be ready - except for the changelog. @luca-heltai can you have a look before I remove the WIP?

bangerth commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,192 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * manifold is later attached to the same triangulation.
+   *
+   * @note For performance and accuracy reasons, it is recommended to apply the
+   * transfinite manifold to as coarse a mesh as possible. Regarding accurcay,
```
typo, btw: "accuracy"

kronbichler commented:
I updated the algorithms considerably: Rather than just taking a single cell that is roughly guessed by the closest distance of points to the unit cell according `MappingQ1::transform_real_to_unit_cell`, I now actually check the transformation and test several cells. This avoids any problems I had before.

I also ran the case from #3874 that previously showed degenerated convergence orders of 3.5 for polynomials of degree 5. When applying a `TransfiniteInterpolationManifold`, I get the following rates:
```
L2 error: 6.28829e-07
L2 slope: 5.97032
L2 error: 1.01813e-08
L2 slope: 5.94867
L2 error: 1.60118e-10
L2 slope: 5.99064
L2 error: 2.64657e-12
L2 slope: 5.91887
L2 error: 7.9474e-13
L2 slope: 1.73557
```
When replacing the `SphericalManifold` with an optimal `PolarManifold`, I get even nicer numbers:
```
L2 error: 3.42185e-08
L2 slope: 6.01173
L2 error: 5.34308e-10
L2 slope: 6.00096
L2 error: 8.34788e-12
L2 slope: 6.00012
L2 error: 1.32365e-13
L2 slope: 5.97881
L2 error: 4.26038e-14
L2 slope: 1.63547
```

luca-heltai commented:
I think this is now in very good shape. I'll take a deeper look in the next hours, but I really like how this is looking! Nice job, Martin! :)

luca-heltai commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * The mechanism to extend the boundary information is a so-called transfinite
+ * interpolation.
+ *
+ * The formula for extending such a description in in 2D is inspired by
```
only one `in`

luca-heltai commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * enough, upon mesh refinement it may happen that the transformation to a
+ * chart inside the get_new_point() or add_new_point() methods produces points
+ * that are outside the unit cells, then this class throws an exception of type
+ * Manifold@<dim,spacedim@>::ExcTransformationFailed. If that happens, the mesh
```
Is this going to happen even with the new implementation?

kronbichler replied:
I have not seen it with the tolerances I have chosen. I used to have a stricter tolerance than `1e-6` in line 806 of `manifold_lib.cc` and then `SphericalManifold` failed. I chose that tolerance since those mappings are quite common in deal.II and I want to at least pass those manifolds. But we will see if the tolerances are appropriate (or should be exposed) when people start using this class.

kronbichler replied:
I must also add that it took me a few days to get this robust and re-wrote the logic two times because one really needs to try the inverse mapping with the given points, it might fail, and one needs to try with another set (usually a neighbor).

luca-heltai commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+  /**
+   * The underlying triangulation.
+   */
+  const Triangulation<dim,spacedim> *triangulation;
```
Shouldn't we be using a `SmartPointer` here?

luca-heltai replied:
or maybe not... This would give us troubles for the fact that `Manifolds` are referenced by `Triangulation`, and closing the loop will always throw an error, no matter who we destroy first, unless we explicitly clear the pointer from the `Triangulation`... just ignore my comment above...

kronbichler replied:
Yes, this is what I was thinking. I will try the signals you suggested, but I'm not sure if I can make it today.

luca-heltai commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+  /**
+   * Initializes the manifold with a coarse mesh. The prerequisite for using
+   * this class is that the input triangulation is uniformly refined and the
+   * manifold is later attached to the same triangulation.
```
Maybe we should warn the user that destroying the underlying `Triangulation` will leave this class in an undefined state. Alternatively, we should attach a `boost::signals2::signal<void ()> Triangulation< dim, spacedim >::Signals::clear` to the given triangulation that resets our pointer to be the `nullptr` when the `Triangulation` is cleared.

kronbichler replied:
I added the signal. Thanks to a lambda it is just a single line of code ;-)

luca-heltai commented on include/deal.II/grid/manifold_lib.h:
```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   * A MappingQ1 object that transforms between the unit and real cell for
+   * idenfifying the correct coarse cell to a given set of surrounding points.
+   */
+  MappingQGeneric<dim,spacedim> mapping_q1;
```
Do we need this one? Wouldn't it be enough to use `StaticMappingQ1<dim,spacedim>::mapping`?

kronbichler replied:
That's true. I will change this.

luca-heltai commented:
I think the only real issue to address is to attach a signal that clears the Triangulation pointer when the underlying triangulation is cleared. After that, squash and merge! Nice job!

kronbichler commented:
A final question before we squash and merge: Is everyone fine with the name? `TransfiniteInterpolationManifold` is a bit verbose, on the other hand it says exactly what happens. I could also imagine `TransfiniteManifold`, but prefer the former.

bangerth commented:
Very nice work! I particularly like the optimal convergence rates!

I'm personally ok with the longer class name. Only those who still use emacs or vi will suffer (and they should switch anyway!), everyone else has autocompletion today.

luca-heltai commented:
Ok with the longer name. Squash and merge! 👍

bangerth commented:
You know, I think you should have put that changelog entry into the "major" category.

kronbichler commented:
Then let's go for that. ;-)

kronbichler commented:
I realize that I had accidentally deleted a block of code during the rebase, causing a large number of test failures [1], essentially all tests using `MappingQGeneric`. Patch in preparation.

[1] https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=7691
